src/wpc/uart_8251.c

```diff
@@ -0,0 +1,138 @@
+
+// If MODE_BAUD(m) != 0, we're in asynchronous mode.
+#define MODE_STOP(m)        ((m >> 6) & 0x03)
+
+// If MODE_BAUD(m) == 0, we're in synchronous mode.
+#define MODE_SYNC_DETECT(m) ((m >> 6) & 0x01)
+#define MODE_SINGLE_SYNC(m) ((m >> 7) & 0x01)
+
+// Macros to parse bits in the command byte sent from the host.
+#define CMD_TX_ENABLE(c)    ((c >> 0) & 0x01)
+#define CMD_DTR_ENABLE(c)   ((c >> 1) & 0x01)
+#define CMD_RX_ENABLE(c)    ((c >> 2) & 0x01)
+#define CMD_SEND_BREAK(c)   ((c >> 3) & 0x01)
+#define CMD_ERROR_RESET(c)  ((c >> 4) & 0x01)
+#define CMD_RTS_ENABLE(c)   ((c >> 5) & 0x01)
+#define CMD_CHIP_RESET(c)   ((c >> 6) & 0x01)
+#define CMD_ENTER_HUNT(c)   ((c >> 7) & 0x01)
+
+
+// Bitmasks for setting the status byte sent to the host
+#define STATUS_TX_READY     (1 >> 0)
+#define STATUS_RX_READY     (1 >> 1)
+#define STATUS_TX_EMPTY     (1 >> 2)
+#define STATUS_PARITY_ERR   (1 >> 3)
+#define STATUS_OVERRUN_ERR  (1 >> 4)
+#define STATUS_FRAME_ERR    (1 >> 5)
+#define STATUS_SYNC_DET     (1 >> 6)
+#define STATUS_DSR          (1 >> 7)
+
+
+static int mode_loaded = 0;
+
+static data8_t mode, command;
+static data8_t rx_byte;
+static data8_t status = (STATUS_DSR | STATUS_TX_READY | STATUS_TX_EMPTY);
+static int baud = 9600;
+static int rx_ready = 0;
+
+data8_t uart_8251_read(int reg)
+{
+    int byte_read;
+
+//    DEBUG_OUT("r 0x%x\n", reg);
+    switch (reg) {
+    case REG_CTRL:
+        // if we don't already have a byte ready, try to read one
+        if (!(status & STATUS_RX_READY)) {
+            byte_read = uart_getch();
+            if (byte_read >= 0) {
+                rx_byte = (data8_t)byte_read;
+                status |= STATUS_RX_READY;
+            }
+        }
+        return status;
+
+    case REG_DATA:
+        // clear the RX_READY bit and return the buffered byte
+        status &= ~STATUS_RX_READY;
+        return rx_byte;
+    }
+
+    // ROM should never read REG_BAUD.  We need a copy of the schematic to
+    // know how actual hardware would respond.
+    return 0x00;
+}
+
+int uart_8251_write(int reg, data8_t value)
+{
+    switch (reg) {
+        case REG_CTRL:
+            if (!mode_loaded) {
+                mode_loaded = 1;
+                mode = value;
+                DEBUG_OUT("mode=0x%02x\n", mode);
+            }
+            else {
+                command = value;
+                if (CMD_CHIP_RESET(command)) {
+                    // next write will set a new mode
+                    mode_loaded = 0;
+                }
+                DEBUG_OUT("cmd=0x%02x\n", command);
+            }
+            break;
+
+        case REG_DATA:
+            uart_putch(value);
+            if (isprint(value) || value == '\r' || value == '\n') {
+                DEBUG_OUT("%c", value);
+            }
+            else {
+                DEBUG_OUT("\nprt 0x%x\n", value);
+            }
+            break;
+
+        case REG_BAUD:
+            baud = 9600 / (value + 1);
```

Comment:
Actually, this should be `9600 / (1 << value)`.  The ROM uses 0 for 9600bps and 5 for 300bps.